### DISCUSSION TITLE
How is luxon's DateTime automatically handled when saving and reading from a database

### DISCUSSION THREAD

**Discussion by vinaysshenoy** [vitaly-t/pg-promise#858] [Q&A] [answered]
Context: I am building a backend application. This is what my stack looks like:

- Language/Platform: Typescript/NodeJS
- Framework: Express (https://expressjs.com/)
- Database: PostgreSQL (https://www.postgresql.org/docs/14/index.html)

Disclaimer: I am not very familiar with the inner workings of JS/TS. I have a working knowledge of the languauge(s).

I use `pg-promise` to connect to my PostgreSQL instance for querying my database. I noticed something interesting when I save and load [`DateTime`](https://moment.github.io/luxon/#/parsing?id=iso-8601) types into my database.

This is the SQL statement I used to create the database:

```sql
CREATE TABLE "PROJECT_TEMPLATES" (
    "id" UUID,
    "name" TEXT NOT NULL,
    "url" TEXT NOT NULL,
    "created_at" TIMESTAMPTZ NOT NULL,
    "updated_at" TIMESTAMPTZ NOT NULL,
    PRIMARY KEY ("id")
);
```

And the code snippet I use to save and query this table:

```typescript
saveProjectTemplate = (project) => {
    const query = `INSERT INTO "PROJECT_TEMPLATES" AS pt
    ("id", "name", "url", "created_at", "updated_at")
    VALUES
    ($(id), $(name), $(url), $(timestamps.createdAt), $(timestamps.updatedAt))
    ON CONFLICT ("id") DO UPDATE
    SET "name" = $(name), "url" = $(url), "updated_at" = $(timestamps.updatedAt)
    WHERE pt."id" = $(id)`;

    return obj.none(query, project);
}

findProjectTemplate = (id) => {
    const query = `SELECT * FROM "PROJECT_TEMPLATES"
    WHERE "id" = $1`;

    return obj.oneOrNone(query, [id]);
}
```

Here, `project` is a `ProjectTemplate` class:

```typescript
class ProjectTemplate {
    public readonly id: string
    public readonly name: string
    public readonly url: string
    public readonly timestamps: Timestamps

    constructor(id: string, name: string, url: string, timestamps: Timestamps) {
        this.id = id
        this.name = name
        this.url = url
        this.timestamps = timestamps
    }
}

class Timestamps {
    public readonly createdAt: DateTime
    public readonly updatedAt: DateTime

    constructor(createdAt: DateTime, updatedAt: DateTime) {
        this.createdAt = createdAt
        this.updatedAt = updatedAt
    }
}
```

And the query returns a type of `ProjectTemplateRecord` which is transformed after querying:

```typescript
interface ProjectTemplateRecord {
    id: string
    name: string
    url: string
    created_at: DateTime
    updated_at: DateTime
}
```

I noticed that the `DateTime` instances get handled automatically and correctly whenever I save and read from the database without the need for me to manually use the `toISO` and `fromISO` methods. How does this work? I just need to know this so I don't get surprised by some behaviour that I am not aware of.

Thanks!

**Note:** I originally posted this on the `Luxon` repository discussions as well, where I could not get a complete answer. See https://github.com/moment/luxon/discussions/1345 for details.

**Comment by vitaly-t** [comment, marked as answer]
Date/Time is formatted according to what's set for [pg-types](https://github.com/brianc/node-pg-types). Check out the defaults there.

Within `pg-promise`, you have access to `pgp.pg.types` to access it:

```ts
import pgPromise from 'pg-promise';

const pgp = pgPromise({/* Initialization Options */});

const {types} = pgp.pg; //=> node-pg-types interface
```

You can tweak all and any formatting handlers via `types`:

```ts
types.setTypeParser(types.builtins.DATE, value => {
    // your custom Date formatter
});
```

**Reply by vinaysshenoy** [reply]
This is very helpful, thank you! I looked at the library's source code and it seems that it eventually parses to JS `Date` objects. Guess I need to understand how Typescript works a bit better since it let me interact with the objects without any errors of the type not matching.

I am assuming that if I had tried to access a method that is available only on the `DateTime` instance and not on `Date`, it would have failed at runtime.

Manually creating a parser now! Thanks for the help!
